server/bleep/src/webserver/answer.rs

```diff
@@ -94,34 +98,141 @@ pub async fn handle(
-        .map_err(|e| {
-            super::error(
-                ErrorKind::Internal,
-                format!("failed to make request to answer API: {}", e),
-            )
-        })?;
-
-    let selection: api::Response = res.json().await.map_err(|e| {
-        super::error(
-            ErrorKind::Internal,
-            format!("answer API was not able to create a valid result: {}", e),
-        )
-    })?;
-
-    let selected_snippet = snippets.remove(selection.data.index as usize);
+        .map_err(|e| super::error(ErrorKind::Internal, e.to_string()))?
+        .answer;
+
+    // reorder snippets
+    let selected_snippet = snippets.remove(res1);
     snippets.insert(0, selected_snippet);
 
     Ok::<_, Json<super::Response<'static>>>(Json(super::Response::Answer(AnswerResponse {
         snippets,
-        selection,
+        selection: api::Response {
+            data: api::DecodedResponse {
+                index: res1 as u32,
+                answer: res2,
+            },
+            id: params.user_id,
+        },
     })))
 }
+
+#[derive(serde::Serialize)]
+struct OpenAIRequest {
+    prompt: String,
+    user_id: String,
+}
+
+struct AnswerAPIClient {
+    client: reqwest::Client,
+    host: String,
+    query: String,
+    user_id: String,
+}
+
+impl AnswerAPIClient {
+    fn new(host: &str, query: &str, user_id: &str) -> Self {
+        Self {
+            client: reqwest::Client::new(),
+            host: host.to_owned(),
+            query: query.to_owned(),
+            user_id: user_id.to_owned(),
+        }
+    }
+
+    async fn send(&self, prompt: String) -> Result<reqwest::Response, reqwest::Error> {
+        self.client
+            .post(self.host.as_str())
+            .json(&OpenAIRequest {
+                prompt: prompt.clone(),
+                user_id: self.user_id.clone(),
+            })
+            .send()
+            .await
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct Stage1Response {
+    index: usize,
+}
+
+#[derive(serde::Deserialize)]
+struct Stage2Response {
+    answer: String,
+}
+
+const DELIMITER: &str = "######";
+impl AnswerAPIClient {
+    async fn stage_1(&self, snippets: &[api::Snippet]) -> Result<Stage1Response, reqwest::Error> {
+        let mut prompt = snippets
+            .iter()
+            .enumerate()
+            .map(|(i, snippet)| {
+                format!(
+                    "Repository: {}\nPath: {}\nLanguage: {}\nIndex: {}\n\n{}\n{DELIMITER}\n",
+                    snippet.repo_name, snippet.relative_path, snippet.lang, i, snippet.text
+                )
+            })
+            .collect::<String>();
+
+        prompt += &format!(
+            "\nAbove are {} code snippets separated by \"{DELIMITER}\". \
+            Your job is to answer which snippet index best answers the question. Your\
+            reply must obey the JSON schema {{ \"index\": number}}.
+
+            Q: What icon do we use to clear search history?
+            A: {{ \"index\": 0 }}
```

Comment:
This'll need to be tested but we don't need the model to return JSON at all. The instruction could be updated to `Your reply should be a single number` and the example could be `A: 0` directly.

Reply:
ah good point, i gave this a go but i was not able to guarantee the output format. i think supplying an example Q/A pair should help with that though.